src/kpro.erl

```diff
@@ -138,764 +213,302 @@
                 | struct_schema()
                 | {array, schema()}
                 | decode_fun(). %% caller defined decoder
--type stack() :: [{tag(), vsn()} | field_name()]. %% encode / decode stack
-
--define(INT, signed-integer).
--define(SCHEMA_MODULE, kpro_schema).
--define(PRELUDE, kpro_prelude_schema).
--define(NO_TIMESTAMP, -1).
+-type stack() :: [{api(), vsn()} | field_name()]. %% encode / decode stack
+-type isolation_level() :: read_committed | read_uncommitted.
+-type connection() :: kpro_connection:connection().
+-type conn_config() :: kpro_connection:config().
+-type vsn_range() :: {vsn(), vsn()}.
+-type vsn_ranges() :: #{api() => vsn_range()}.
+-type protocol() :: plaintext | ssl | sasl_plaintext | sasl_ssl.
+-type coordinator_type() :: group | txn.
+-type group_id() :: binary().
+-type transactional_id() :: binary().
+-type producer_epoch() :: int16().
+-type txn_ctx() :: #{ connection => connection()
+                    , transactional_id => transactional_id()
+                    , producer_id => producer_id()
+                    , producer_epoch => producer_id()
+                    }.
+-type batch_meta_key() ::
+        is_transaction % if this message was produced in a transaction
+      | is_control % for read_uncommitted clients
+      | last_offset % client wont have to do lists:last(Messages)
+      | max_ts % client don't have to scan the messages for max
+      | producer_id. % it can be referenced by a future fetch response in
+                     % its aborted_transactions field
+-type batch_meta_val() :: boolean() | atom() | integer().
+-type batch_meta() :: ?KPRO_NO_BATCH_META %% magic 0-1
+                    | #{batch_meta_key() => batch_meta_val()}.
+-type batch_decode_result() :: ?incomplete_batch(int32())
+                             | {batch_meta(), [message()]}.
+%% offset or offset + associated user-data to commit
+-type offset_ud() :: offset() %% no user data
+                   | {offset(), binary()}.
+-type offsets_to_commit() :: #{{topic(), partition()} => offset_ud()}
+                           | [{{topic(), partition()}, offset_ud()}].
+
+%% All versions of kafka messages (records) share the same header:
+%% Offset => int64
+%% Length => int32
+%% We need to at least fetch 12 bytes in order to fetch:
+%%  - one complete message when it's magic v0-1 not compressed
```

Comment:
Nit: s/"it's"/its/g (also 1 line below)

Reply:
It meant to mean "when it is a magic version 0-1 non-compressed batch".
will reword.